d3a_api_client/websocket_device.py

```diff
@@ -0,0 +1,80 @@
+import asyncio
+import logging
+import json
+import threading
+import traceback
+import websockets
+from d3a_interface.utils import wait_until_timeout_blocking
+
+
+class RestWebsocketAPIException(Exception):
+    pass
+
+
+class WebsocketMessageDispatcher:
+    def __init__(self, rest_client):
+        self.client = rest_client
+        self.command_response_buffer = []
+
+    def received_message(self, message):
+        if "event" in message:
+            if message["event"] == "market":
+                self.client._on_market_cycle(message)
+            elif message["event"] == "tick":
+                self.client._on_tick(message)
+            elif message["event"] == "trade":
+                self.client._on_trade(message)
+            else:
+                logging.error(f"Received message with unknown event type: {message}")
+        elif "command" in message:
+            self.command_response_buffer.append(message)
+
+    def wait_for_command_response(self, command_name):
+        def check_if_command_response_received():
+            return any("command" in c and c["command"] == command_name
+                       for c in self.command_response_buffer)
+
+        logging.info(f"Command {command_name} waiting for response...")
+        wait_until_timeout_blocking(check_if_command_response_received, timeout=120)
+        response = next(c
+                        for c in self.command_response_buffer
+                        if "command" in c and c["command"] == command_name)
+        self.command_response_buffer.remove(response)
+        return response
+
+
+async def websocket_coroutine(websocket_uri, websocket_headers, message_dispatcher):
+    async with websockets.connect(websocket_uri, extra_headers=websocket_headers) as websocket:
+        while True:
```

Comment:
I wonder why you need a `while True` here shouldnt the async def and assync with deal with reiterating?

Reply:
All `await` does is to wait asynchronously until the command returns. Apart from that, not async nor await have any iteration logic. What has iteration logic is the event loop, but cannot be used this way to my knowledge (by default the event loop is processing events that are market as async, and cannot force a function to be triggered multiple times)

Reply:
ok, thanks for the explanation!

Reply:
More on that on the workshop next week!